pkg/provider/provider.go

```diff
@@ -31,17 +32,18 @@ type Responsibility interface {
 }
 
 // Provider represents a bundle of responsibilities defined by a specific ID.
-// Examples of providers are the heavy client, the storagenode, and the gateway.
+// Examples of providers are the satellite, the storagenode, and the uplink.
 type Provider struct {
 	lis      net.Listener
 	grpc     *grpc.Server
 	next     []Responsibility
 	identity *FullIdentity
+	nodeType pb.NodeType
 }
 
 // NewProvider creates a Provider out of an Identity, a net.Listener, a UnaryInterceptorProvider and
 // a set of responsibilities.
-func NewProvider(opts *ServerOptions, lis net.Listener, interceptor grpc.UnaryServerInterceptor,
+func NewProvider(opts *ServerOptions, lis net.Listener, interceptor grpc.UnaryServerInterceptor, nodeType pb.NodeType,
```

Comment:
we can actually get away without changing the provider at all!

```
// StorageNodeConfig is a Config that implements provider.Responsibility as
// a storage node
type StorageNodeConfig Config

// Run implements provider.Responsibility
func (c StorageNodeConfig) Run(ctx context.Context, server *provider.Provider) error {
	return Config(c).Run(ctx, server, pb.NodeType_STORAGE)
}

// SatelliteConfig is a Config that implements provider.Responsibility as
// a satellite
type SatelliteConfig Config

// Run implements provider.Responsibility
func (c SatelliteConfig) Run(ctx context.Context, server *provider.Provider) error {
	return Config(c).Run(ctx, server, pb.NodeType_SATELLITE)
}

// Run does not implement provider.Responsibility. Please use a specific
// SatelliteConfig or StorageNodeConfig
func (c Config) Run(ctx context.Context, server *provider.Provider,
	nodeType pb.NodeType) (err error) {
```

then all you have to do is update commands to use kademlia.StorageNodeConfig or kademlia.SatelliteConfig instead of kademlia.Config, and everything in pkg/provider can be left alone

Reply:
This is nice. I forgot about this approach.